src/cargo/core/compiler/build_context/target_info.rs

```diff
@@ -694,13 +689,44 @@ impl RustcTargetData {
             target_config.insert(ct, host_config.clone());
         }
 
-        Ok(RustcTargetData {
+        let mut res = RustcTargetData {
             rustc,
+            config,
+            requested_kinds: requested_kinds.into(),
             target_config,
             target_info,
             host_config,
             host_info,
-        })
+        };
+
+        // Get all kinds we currently know about. Further kinds will
+        // be added via the `add_kind` function below.
+        let all_kinds = requested_kinds
+            .iter()
+            .copied()
+            .chain(ws.members().flat_map(|p| {
+                p.manifest()
+                    .default_kind()
+                    .into_iter()
+                    .chain(p.manifest().forced_kind())
+            }));
+        for kind in all_kinds {
+            res.add_kind(&kind)?;
+        }
+
+        Ok(res)
+    }
+
+    pub fn add_kind(&mut self, kind: &CompileKind) -> CargoResult<()> {
+        if let CompileKind::Target(target) = kind {
+            let tcfg = self.config.target_cfg_triple(target.short_name())?;
+            self.target_config.insert(*target, tcfg);
+            self.target_info.insert(
+                *target,
+                TargetInfo::new(self.config, &self.requested_kinds, &self.rustc, *kind)?,
+            );
```

Comment:
Since this might presumably be called for the same `target` multiple times could this avoid doing work if we alread know about `target`?